File: static/js/world-control.js
/* ══ THE COMPANY — Control Plane ══
   Mission control: one master switch for the whole Company, a toggle for every
   autonomy system, and trigger buttons for the Company's capabilities (its
   "mini-MCP"). Uses _worldModal + api()/esc()/toast(). Global-scope classic script. */

async function worldControl() {
  let d;
  try { d = await api('/api/world/control/panel'); }
  catch (e) { toast?.('Control panel failed to load'); return; }
  _renderControl(d);
}

/* Plain-language help for each autonomy toggle: what it turns on downstream. */
const _SYS_HELP = {
  create: 'Turns on the Studio making art/music/video/3D by itself on the GPU box; finished pieces land in the queue and in Review. Runs only while the master switch is on.',
  govern: 'Lets the Republic (the agent assembly) convene itself every few hours to set strategy and take action. Gated by the master switch.',
  cognition: 'Agents form thoughts, opinions and ideas via the local LLM on the GPU box. Free - no real money involved.',
  meetings: 'Periodic town-hall votes where the agents decide company direction.',
  incidents: 'Injects random in-world events to keep the simulation lively. Cosmetic - no real-world effect.',
  auto_publish: 'Sets the God Console mode. On = free publishes (example.com WordPress, Cults3D) run on their own; paid Etsy/Printify listings, PayPal payouts and code changes ALWAYS still wait for your blessing.',
  sell: 'Lets agents draft paid Etsy/Printify listings by themselves. Each one queues as a prayer in the God Console for you to approve (about 0.20 dollars per Etsy listing from the treasury).',
  swarm_cron: 'Advances cron-scheduled coding jobs in the Dev-Swarm on schedule.',
  sec_monitor: 'Watches the home network for new or changed devices.',
  sec_scan: 'Runs periodic Pi-hole and network config scans.',
  sec_analyze: 'Runs LLM threat analysis on the GPU box and flags suspicious activity.',
  sec_audit: 'Nightly hardening snapshot; any regression alerts you in the God Console.',
  guardian: 'Auto-blocks ad, tracker and ACR domains network-wide - never local or functional ones. Fully reversible.',
  ai_watch: 'Watches the AI agents themselves for rogue behaviour (payout or code bursts, unknown actors) and reports to the God Console.',
};

/* What each Capability trigger actually does when clicked. */
const _CAP_HELP = {
  make_art: 'Runs one Studio image generation now on the GPU box; the result lands in the queue and Review.',
  make_music: 'Runs one Studio music generation now on the GPU box; the result lands in the queue and Review.',
  make_video: 'Runs one Studio video generation now on the GPU box; the result lands in the queue and Review.',
  make_3d: 'Sculpts one 3D model now on the GPU box; the result lands in the queue and Review.',
  sell_etsy: 'Drafts an Etsy listing now; it appears as a prayer in the God Console for you to approve before it goes live (about 0.20 dollars).',
  sell_printify: 'Drafts a Printify listing now; it appears as a prayer in the God Console for you to approve before it goes live.',
  convene: 'Convenes the Republic assembly right now to strategize and act.',
  research: 'Asks you for a topic, then files a research request as a prayer in the God Console.',
  scan_trends: 'Scans market trends right now to surface product ideas.',
  sec_scan_now: 'Runs a one-off security scan of the network right now.',
};

function _pill(on, disabled) {
  const bg = on ? '#1f4a32' : '#26324a', dot = on ? 'right:2px' : 'left:2px', col = on ? '#6ee7a8' : '#54607a';
  return `<span style="display:inline-block;position:relative;width:38px;height:20px;border-radius:11px;background:${bg};border:1px solid ${col}44;vertical-align:middle;${disabled ? 'opacity:.5' : ''}">
    <span style="position:absolute;top:2px;${dot};width:14px;height:14px;border-radius:50%;background:${col};transition:all .12s"></span></span>`;
}

function _renderControl(d) {
  const m = d.master;

  const master = `
    <div style="display:flex;justify-content:space-between;align-items:center;gap:10px;padding:14px 16px;margin-bottom:16px;border-radius:12px;
      background:${m ? 'linear-gradient(135deg,#12251b,#0f1626)' : '#1a1414'};border:1px solid ${m ? '#2a5a3a' : '#5a2a2a'}">
      <div>
        <div style="font-weight:800;font-size:1.05rem;color:#e8eefc">${m ? '🟢 The Company is running' : '⏸️ The Company is dormant'}</div>
        <div style="font-size:.74rem;color:#8a97ad;margin-top:2px">Master switch — gates every autonomous system below. ${hlp('When off, every system below is forced off on its next tick regardless of its own switch. A system is effective only when this master AND its own toggle are on.')}</div>
      </div>
      <button class="btn" style="padding:8px 16px;font-weight:700;${m ? 'background:#5a2a2a;border-color:#7c3a3a' : 'background:#1f4a32;border-color:#2a5a3a;color:#6ee7a8'}" onclick="controlMaster(${m ? 'false' : 'true'})">${m ? '⏸ Pause all' : '▶ Wake the Company'}</button>
    </div>`;

  // run mode: how fast + how autonomous the sim runs
  const rm = d.run_mode || 'normal';
  const rmBtn = (id, label, desc) => `<button class="btn" title="${esc(desc)}" onclick="controlRunMode('${id}')"
    style="flex:1;padding:8px 6px;font-weight:600;${rm === id ? 'background:#1f3a5a;border-color:#3a6a9a;color:#8ecaff' : ''}">${label}</button>`;
  const runmode = `
    <div style="padding:10px 14px;margin-bottom:14px;border-radius:12px;background:#0e1626;border:1px solid #26324a">
      <div style="font-size:.74rem;color:#9fc0ff;font-weight:600;margin-bottom:6px">⏱️ Run mode <span style="color:#8a97ad;font-weight:400">· how fast + how autonomous the sim runs</span></div>
      <div style="display:flex;gap:6px">
        ${rmBtn('normal', '🐢 Normal', 'Real pace, your automation settings as-is.')}
        ${rmBtn('fast', '⚡ Fast', '~5x sim speed — watch the town work, gather and advance. LLM/GPU/money keep their own real cadences + gates.')}
        ${rmBtn('test', '🧪 Test', 'Fast + auto-run the FREE loops (art/work/era) so the world visibly progresses. Money & code stay gated — nothing real is spent or posted.')}
      </div>
    </div>`;

  // systems grouped
  const groups = {};
  (d.systems || []).forEach(s => (groups[s.group] = groups[s.group] || []).push(s));
  const sysHtml = Object.entries(groups).map(([g, arr]) => `
    <div style="font-size:.66rem;color:#7a86a0;text-transform:uppercase;letter-spacing:.06em;margin:10px 0 4px">${esc(g)}</div>
    ${arr.map(s => {
      const paused = m && s.desired && !s.effective;   // shouldn't happen, but safe
      const gated = s.desired && !m;
      return `<div style="display:flex;justify-content:space-between;align-items:center;gap:10px;padding:8px 10px;margin-bottom:5px;background:#0e1626;border:1px solid #26324a;border-radius:8px;cursor:pointer" onclick="controlSystem('${s.id}',${s.desired ? 'false' : 'true'})">
        <div style="min-width:0">
          <div style="font-size:.82rem;color:#e8eefc">${esc(s.label)} ${_SYS_HELP[s.id] ? hlp(_SYS_HELP[s.id]) : ''} ${gated ? '<span style="font-size:.64rem;color:#f59e0b">· paused by master</span>' : ''}</div>
          <div style="font-size:.68rem;color:#7a86a0">${esc(s.desc || '')}</div>
        </div>
        ${_pill(s.desired, false)}
      </div>`;
    }).join('')}`).join('');

  // capabilities grouped
  const capG = {};
  (d.capabilities || []).forEach(c => (capG[c.group] = capG[c.group] || []).push(c));
  const capHtml = `
    <div style="font-weight:700;color:#e8eefc;margin:16px 0 6px">⚡ Capabilities <span style="font-size:.66rem;color:#7a86a0;font-weight:400">— trigger an action, get a product back</span></div>
    ${Object.entries(capG).map(([g, arr]) => `
      <div style="display:flex;gap:6px;flex-wrap:wrap;align-items:center;margin-bottom:6px">
        <span style="font-size:.66rem;color:#54607a;width:56px">${esc(g)}</span>
        ${arr.map(c => `<button class="btn" style="padding:4px 10px;font-size:.74rem"${_CAP_HELP[c.id] ? ` title='${esc(_CAP_HELP[c.id])}'` : ''} onclick="controlTrigger('${c.id}')">${esc(c.label)}</button>`).join('')}
      </div>`).join('')}`;

  // selling config
  const sl = d.sell || {};
  const sellHtml = `
    <div style="font-weight:700;color:#e8eefc;margin:16px 0 6px">💰 Selling <span style="font-size:.66rem;color:#7a86a0;font-weight:400">— pricing for paid listings (always reviewed before going live)</span></div>
    <div style="display:flex;gap:8px;flex-wrap:wrap;align-items:center;padding:10px;background:#0e1626;border:1px solid #26324a;border-radius:8px">
      <span style="font-size:.72rem;color:#8a97ad">Price ${hlp('The list price agents put on new paid Etsy/Printify listings. Every listing is still reviewed as a prayer before it goes live.')} $<input id="sell-price" type="number" step="0.01" value="${((sl.price_cents || 2499) / 100).toFixed(2)}" style="width:74px;padding:4px 7px;background:#0b1120;border:1px solid #26324a;border-radius:6px;color:#e8eefc"></span>
      <span style="font-size:.72rem;color:#8a97ad">Sold as ${hlp('What the design is sold as (e.g. Poster, T-Shirt). Sets the product type on new Etsy/Printify listings.')} <input id="sell-ptype" value="${esc(sl.product_type || 'Poster')}" style="width:100px;padding:4px 7px;background:#0b1120;border:1px solid #26324a;border-radius:6px;color:#e8eefc"></span>
      <button class="btn" style="padding:4px 10px;font-size:.72rem" onclick="controlSellSave()">💾 Save</button>
      <span style="font-size:.68rem;color:#54607a;width:100%">
        Etsy: <b style="color:${sl.etsy_ready ? '#6ee7a8' : '#f59e0b'}">${sl.etsy_ready ? 'connected ✓' : 'not connected'}</b> ·
        Printify: <b style="color:${sl.printify_ready ? '#6ee7a8' : '#f59e0b'}">${sl.printify_ready ? 'connected ✓' : 'add API key'}</b> ·
        each Etsy listing costs $0.20 from the treasury.
      </span>
    </div>`;

  _worldModal('🎛️ Company Control', master + runmode + sysHtml + capHtml + sellHtml);
}

async function controlSellSave() {
  const price = parseFloat(document.getElementById('sell-price')?.value || '24.99');
  const product_type = document.getElementById('sell-ptype')?.value || 'Poster';
  try { const d = await api('/api/world/control/sell-config', { method: 'POST', body: JSON.stringify({ price_dollars: price, product_type }) }); toast?.('Saved'); _renderControl(d); }
  catch (e) { toast?.('Failed'); }
}

async function controlMaster(on) {
  try { const d = await api('/api/world/control/master', { method: 'POST', body: JSON.stringify({ on }) }); toast?.(on ? '🟢 Company running' : '⏸️ Company paused'); _renderControl(d); }
  catch (e) { toast?.('Failed'); }
}
async function controlSystem(id, on) {
  try { const d = await api('/api/world/control/system', { method: 'POST', body: JSON.stringify({ id, on }) }); _renderControl(d); }
  catch (e) { toast?.(e?.message || 'Failed'); }
}
async function controlTrigger(id) {
  let args = {};
  if (id === 'research') {
    const topic = prompt('What should the scholars research?');
    if (topic == null) return;
    args = { topic };
  }
  try {
    const r = await api('/api/world/control/trigger', { method: 'POST', body: JSON.stringify({ id, args }) });
    toast?.(r.ok ? `⚡ ${r.result || 'triggered'}` : (r.error || 'Failed'));
  } catch (e) { toast?.('Failed'); }
}

window.worldControl = worldControl;
async function controlRunMode(mode) {
  try {
    await api('/api/world/control/runmode', { method: 'POST', body: JSON.stringify({ mode }) });
    toast?.(mode === 'test' ? '🧪 Test — dry run, evolving fast (money/code still gated)'
          : mode === 'fast' ? '⚡ Fast — ~5x sim speed'
          : '🐢 Normal pace');
    const d = await api('/api/world/control/panel'); _renderControl(d);
  } catch (e) { toast?.(e.message); }
}
window.controlRunMode = controlRunMode;
window.controlMaster = controlMaster;
window.controlSystem = controlSystem;
window.controlTrigger = controlTrigger;
window.controlSellSave = controlSellSave;

File: app/world_control.py
"""
The Company — unified automation control plane.

One place the Company owns ALL automation. Two kinds of thing:

  • Autonomy systems — background loops that run on their own (creation, the
    Republic's self-governance, the crew thinking, town meetings, dev-swarm cron,
    security scans). Each has a toggle; a single MASTER switch gates them all.

  • Capabilities ("mini-MCP") — discrete actions the Company can TRIGGER on
    demand and receive a product back (make art/music/video/3D, convene the
    assembly, commission research, scan trends). No autonomy to manage — just
    fire and receive.

Design is deliberately NON-INVASIVE: every autonomy system already reads its own
native setting each tick. We don't rewrite those loops — the master + per-system
toggles simply CASCADE into the native settings (effective = master AND desired).
So flipping the master off writes every native switch off, and each existing
loop stops on its next tick. One tiny exception (dev-swarm cron) gets a global
gate added in scheduler.py.
"""
import json, logging, threading
import httpx
from deps import get_conn, get_setting

logger = logging.getLogger("store")
_LOCAL = "http://127.0.0.1:8787"


# ── settings io ──────────────────────────────────────────────────────────────
def _get(k, d=None):
    return get_setting(k, d)


def _set(k, v):
    c = get_conn()
    try:
        c.execute("INSERT OR REPLACE INTO settings (key,value) VALUES (?,?)", (k, str(v)))
        c.commit()
    finally:
        c.close()


def _truthy(v):
    return str(v).lower() in ("1", "true", "yes", "on")


# ── autonomy systems: id → native setting it cascades into ───────────────────
SYSTEMS = [
    {"id": "create",      "label": "Autonomous creation",   "group": "Company", "key": "world_auto_enabled",         "on": "1",   "off": "0", "desc": "The studio makes art/music/video/3D on a cadence."},
    {"id": "govern",      "label": "Republic self-governs",  "group": "Company", "key": "world_auto_govern_min",      "on": "360", "off": "0", "numeric": True, "desc": "The assembly convenes itself to strategize + act."},
    {"id": "cognition",   "label": "Crew thinking (LLM)",    "group": "World",   "key": "world_llm_enabled",          "on": "1",   "off": "0", "desc": "Agents form thoughts, opinions, ideas."},
    {"id": "meetings",    "label": "Town meetings",          "group": "World",   "key": "world_meetings_enabled",     "on": "1",   "off": "0", "desc": "Periodic votes on company direction."},
    {"id": "incidents",   "label": "Incidents & events",     "group": "World",   "key": "world_incidents_enabled",    "on": "1",   "off": "0", "desc": "Random world events keep it lively."},
    {"id": "auto_publish", "label": "Auto-publish (free only)", "group": "Company", "key": "world_ops_automation_mode", "on": "budget", "off": "review", "literal": True, "desc": "Free publishes (WordPress, Cults3D) run on their own. Paid listings, payouts and code ALWAYS wait for your blessing."},
    {"id": "sell",        "label": "Autonomous listing",     "group": "Company", "key": "world_sell_auto",            "on": "1",   "off": "0", "desc": "Queue paid Etsy/Printify listings for your review (costs $0.20/ea)."},
    {"id": "swarm_cron",  "label": "Dev-swarm cron",         "group": "Dev",     "key": "swarm_cron_enabled",         "on": "1",   "off": "0", "desc": "Advance cron-enabled coding jobs on schedule."},
    {"id": "sec_monitor", "label": "Security monitor",       "group": "Infra",   "key": "security_monitor_enabled",   "on": "1",   "off": "0", "desc": "Watch the network for device changes."},
    {"id": "sec_scan",    "label": "Security auto-scan",     "group": "Infra",   "key": "security_autoscan_enabled",  "on": "1",   "off": "0", "desc": "Periodic Pi-hole / config scans."},
    {"id": "sec_analyze", "label": "AI threat-hunt",         "group": "Infra",   "key": "security_autoanalyze_enabled", "on": "1", "off": "0", "desc": "LLM threat analysis (uses the GPU)."},
    {"id": "sec_audit",   "label": "Nightly security audit", "group": "Infra",   "key": "security_audit_enabled",     "on": "1",   "off": "0", "desc": "Hardening snapshot + regression alerts to the God Console."},
    {"id": "guardian",    "label": "Network Guardian (auto-block)", "group": "Infra", "key": "netguard_auto_enabled",  "on": "1",   "off": "0", "desc": "Auto-block ad/tracker/ACR domains network-wide — never functional/local. Reversible."},
    {"id": "ai_watch",    "label": "AI Shield (agent watch)",   "group": "Infra",   "key": "ai_watch_enabled",           "on": "1",   "off": "0", "desc": "Watch AI agents for rogue behaviour (payout/code bursts, unknown actors) → God Console."},
]
_SYS = {s["id"]: s for s in SYSTEMS}


def _val_on(s, raw):
    """Interpret a raw setting value as on/off for this system's type."""
    if s.get("literal"):                 # non-boolean setting: on means raw == the "on" value
        return raw == s["on"]
    if s.get("numeric"):
        try:
            return int(raw) > 0
        except Exception:
            return False
    return _truthy(raw)


def _native_on(s):
    return _val_on(s, _get(s["key"], s["off"]))


def master_on():
    return _truthy(_get("company_master_on", "1"))


def desired(sid):
    """Per-system INTENT = the native setting itself — the SINGLE SOURCE OF TRUTH. Control
    no longer keeps a `company_desired_*` shadow that got re-cascaded over the native keys
    (on every import + every panel action), silently reverting edits you made directly in
    the ⚙️ Settings modal / 🛡️ Security Command. While the master is PAUSED all natives are
    forced off, so we surface the saved pre-pause value so the panel shows what will resume."""
    s = _SYS[sid]
    if not master_on():
        saved = _get(f"company_paused_{s['key']}", None)
        if saved is not None:
            return _val_on(s, saved)
    return _native_on(s)


def set_master(on):
    """Master = a non-destructive PAUSE. Pausing snapshots each system's current native
    value then forces it off; resuming restores from the snapshot. This never clobbers a
    value you set directly elsewhere (unlike the old cascade)."""
    was = master_on()
    if on and not was:                                    # RESUME → restore pre-pause states
        for s in SYSTEMS:
            saved = _get(f"company_paused_{s['key']}", None)
            if saved is not None:
                _set(s["key"], saved)
        _set("company_master_on", "1")
    elif (not on) and was:                                # PAUSE → snapshot, then force off
        for s in SYSTEMS:
            _set(f"company_paused_{s['key']}", _get(s["key"], s["off"]))
            _set(s["key"], s["off"])
        _set("company_master_on", "0")
    else:
        _set("company_master_on", "1" if on else "0")
    try:
        import world_ops as wo
        wo.note("🟢 The Company is awake — automation resumed." if on
                else "⏸️ The Company is dormant — all automation paused.",
                kind="info", from_agent="Mission Control")
    except Exception:
        pass
    return panel()


def set_system(sid, on):
    """Write the NATIVE key directly (the source of truth). While paused, remember the
    intent in the snapshot so it takes effect when the master resumes."""
    if sid not in _SYS:
        return None
    s = _SYS[sid]
    val = s["on"] if on else s["off"]
    if master_on():
        _set(s["key"], val)
    else:
        _set(f"company_paused_{s['key']}", val)
    return panel()


def init():
    """First-run ONLY: set the full-auto baseline once, natively. NEVER cascades on later
    imports — that silent re-apply was the bug that reverted native edits every restart."""
    if _get("company_control_init") == "1":
        return
    full_auto = {"create", "govern", "cognition", "meetings", "incidents"}
    for s in SYSTEMS:
        if s["id"] in full_auto:
            _set(s["key"], s["on"])
    _set("company_master_on", "1")
    _set("company_control_init", "1")
    logger.info("world_control: first-run full-auto baseline set (native keys)")


# ── capabilities ("mini-MCP"): trigger an action, receive a product ──────────
def _cap_create(kind):
    import world_auto
    if world_auto._state["running"]:
        return "a creation is already in progress"
    threading.Thread(target=world_auto.run_cycle, args=(kind, True), daemon=True).start()
    return f"creating {kind}"


def _cap_convene():
    import world_strategy
    threading.Thread(target=world_strategy.run_cycle, daemon=True).start()
    return "the assembly is convening"


def _cap_research(args):
    import world_ops as wo
    topic = (args or {}).get("topic") or "Scout a survival edge on the web"
    wo.pray("library_research", f"Research: {topic}", detail="Commissioned from Mission Control.",
            cost_cents=0, agent_name="Mission Control")
    return f"research commissioned: {topic}"


def _cap_sell(channel):
    import world_sell
    threading.Thread(target=world_sell.list_design, args=(channel,), daemon=True).start()
    return f"drafting a {channel} listing — it'll appear as a prayer to review"


def _cap_http(method, path):
    def go():
        try:
            httpx.request(method, f"{_LOCAL}{path}", timeout=20)
        except Exception:
            logger.exception("capability http %s %s", method, path)
    threading.Thread(target=go, daemon=True).start()
    return "triggered"


CAPABILITIES = [
    {"id": "make_art",    "label": "🖼️ Create art",           "group": "Create", "fn": lambda a: _cap_create("image")},
    {"id": "make_music",  "label": "🎵 Create music",          "group": "Create", "fn": lambda a: _cap_create("music")},
    {"id": "make_video",  "label": "🎬 Create video",          "group": "Create", "fn": lambda a: _cap_create("video")},
    {"id": "make_3d",     "label": "🧊 Sculpt a 3D model",     "group": "Create", "fn": lambda a: _cap_create("3d")},
    {"id": "sell_etsy",   "label": "🛍️ List a design on Etsy", "group": "Sell",   "fn": lambda a: _cap_sell("etsy")},
    {"id": "sell_printify", "label": "👕 List on Printify",     "group": "Sell",   "fn": lambda a: _cap_sell("printify")},
    {"id": "convene",     "label": "🏛️ Convene the assembly",  "group": "Govern", "fn": lambda a: _cap_convene()},
    {"id": "research",    "label": "📖 Commission research",   "group": "Learn",  "fn": lambda a: _cap_research(a)},
    {"id": "scan_trends", "label": "📈 Scan market trends",    "group": "Research", "fn": lambda a: _cap_http("POST", "/api/trends/scan")},
    {"id": "sec_scan_now", "label": "🛡️ Run a security scan",  "group": "Infra",  "fn": lambda a: _cap_http("POST", "/api/security/scan")},
]
_CAP = {c["id"]: c for c in CAPABILITIES}


def invoke(cap_id, args=None):
    c = _CAP.get(cap_id)
    if not c:
        return {"ok": False, "error": f"unknown capability '{cap_id}'"}
    try:
        return {"ok": True, "result": c["fn"](args or {})}
    except Exception as e:
        logger.exception("capability %s failed", cap_id)
        return {"ok": False, "error": str(e)}


def set_sell_config(price_cents=None, product_type=None):
    if price_cents is not None:
        _set("world_sell_price_cents", max(100, int(price_cents)))
    if product_type:
        _set("world_sell_product_type", str(product_type)[:40])
    return panel()


def panel():
    m = master_on()
    try:
        price = int(_get("world_sell_price_cents", "2499"))
    except Exception:
        price = 2499
    return {
        "master": m,
        "run_mode": _get("world_run_mode", "normal"),
        "systems": [{
            "id": s["id"], "label": s["label"], "group": s["group"], "desc": s.get("desc", ""),
            "desired": desired(s["id"]), "effective": (m and _native_on(s)),
        } for s in SYSTEMS],
        "capabilities": [{"id": c["id"], "label": c["label"], "group": c["group"]} for c in CAPABILITIES],
        "sell": {
            "price_cents": price,
            "product_type": _get("world_sell_product_type", "Poster"),
            "etsy_ready": bool(_get("etsy_access_token") and _get("etsy_shop_id")),
            "printify_ready": bool(_get("printify_key") and _get("printify_shop_id")),
        },
    }


init()
